libsodium-sys/build.rs

```diff
@@ -1,38 +1,91 @@
+extern crate zip;
 
+use http_req::request;
+use sha2::{Digest, Sha256};
 use std::env;
-use std::path::PathBuf;
+use std::fs;
+use std::io::Cursor;
+use std::path::{Path, PathBuf};
 
-#[cfg(target_env = "msvc")]
-extern crate vcpkg;
+static DOWNLOAD_BASE_URL: &'static str = "https://download.libsodium.org/libsodium/releases/";
+static VERSION: &'static str = "1.0.16";
+
+#[cfg(target_env = "msvc")] // libsodium-<VERSION>-msvc.zip
+static SHA256: &'static str = "0580d54f57594a7cb493607cec6e7045369fb67d43623491523781e901589948";
+
+#[cfg(all(windows, not(target_env = "msvc")))] // libsodium-<VERSION>-mingw.tar.gz
+static SHA256: &'static str = "5b81a4fc5d0de36dbda7efeaf355c133d4f6cc0b4dbf69bbe46ef7f5a6baa639";
+
+#[cfg(not(windows))] // libsodium-<VERSION>.tar.gz
+static SHA256: &'static str = "eeadc7e1e1bcef09680fb4837d448fbdf57224978f865ac1c16745868fbd0533";
 
 fn main() {
     println!("cargo:rerun-if-env-changed=SODIUM_LIB_DIR");
     println!("cargo:rerun-if-env-changed=SODIUM_INC_DIR");
     println!("cargo:rerun-if-env-changed=SODIUM_STATIC");
+    println!("cargo:rerun-if-env-changed=SODIUM_FROM_SRC");
+
     if cfg!(target_env = "msvc") {
         // vcpkg requires to set env VCPKGRS_DYNAMIC
         println!("cargo:rerun-if-env-changed=VCPKGRS_DYNAMIC");
     }
 
+    let from_src = env::var("SODIUM_FROM_SRC").is_ok();
     let include_dir = {
-        let lib_dir_isset = env::var_os("SODIUM_LIB_DIR").is_some();
-        let inc_dir_isset = env::var_os("SODIUM_INC_DIR").is_some();
-        if lib_dir_isset || inc_dir_isset {
-            find_libsodium_env()
+        if from_src {
+            Some(build_libsodium())
         } else {
-            find_libsodium_pkg()
+            let lib_dir_isset = env::var_os("SODIUM_LIB_DIR").is_some();
+            let inc_dir_isset = env::var_os("SODIUM_INC_DIR").is_some();
+            if lib_dir_isset || inc_dir_isset {
+                if from_src {
+                    panic!("SODIUM_FROM_SRC is incompatible with SODIUM_LIB_DIR and SODIUM_INC_DIR");
+                }
+                find_libsodium_env()
+            } else {
+                match find_libsodium_pkg() {
+                    Some(lib) => Some(lib),
+                    None => Some(build_libsodium()),
+                }
+            }
         }
     };
 
+    let mode = match env::var_os("SODIUM_STATIC") {
+        Some(_) => "static",
+        None => "dylib",
+    };
+
+    if cfg!(target_env = "msvc") {
+        println!("cargo:rustc-link-lib={0}=libsodium", mode);
+    } else {
+        println!("cargo:rustc-link-lib={0}=sodium", mode);
+    }
+
     let bindings = bindgen::Builder::default()
         .header("sodium_wrapper.h")
         .whitelist_function("(sodium|crypto|randombytes)_.*")
         .whitelist_type("(sodium|crypto|randombytes)_.*")
         .whitelist_var("(sodium|crypto|randombytes)_.*")
         .clang_arg("-isystem")
-        .clang_arg(include_dir)
+        .clang_arg(include_dir.unwrap())
```

Comment:
@Dylan-DPC @kurnevsky and I came to a conclusion that we must get rid of bindgen dependency since libsodium .h files do not have any platform-dependent code.

Since there are issues like that https://github.com/sodiumoxide/sodiumoxide/issues/280 and https://github.com/sodiumoxide/sodiumoxide/issues/267 and https://github.com/sodiumoxide/sodiumoxide/issues/231 the best option is to use the same scenario as rust_sodium does: https://github.com/maidsafe/rust_sodium/blob/master/rust_sodium-sys/build.rs#L64

Reply:
So you have to adopt their script https://github.com/maidsafe/rust_sodium/blob/master/rust_sodium-sys/src/lib.rs#L32 and import bindinds.rs into the repo. No need to regen them each build.

Reply:
feels like this should be a separate effort, unless I'm missing something

Reply:
Yes. We can do that if you don't have time.

Reply:
That's the whole point. See https://github.com/sodiumoxide/sodiumoxide/issues/280

Reply:
Ok, but that means that static linking with downloaded libsodium should be default behavior. And perhaps build with different versions shouldn't be allowed. Otherwise everyone will use system libsodium ignoring it's version. I'm fine with statically generated bindings if we make these changes.